platform/darwin/test/MGLExpressionTests.mm

```diff
@@ -0,0 +1,176 @@
+#import <XCTest/XCTest.h>
+
+#import <string>
+
+#import "NSExpression+MGLAdditions.h"
+
+@interface MGLExpressionTests : XCTestCase
+
+@end
+
+@implementation MGLExpressionTests
+
+#pragma mark - Utility
+
+- (NSComparisonPredicate *)equalityComparisonPredicateWithRightConstantValue:(id)rightConstantValue
+{
+    NSComparisonPredicate *predicate = [NSComparisonPredicate
+        predicateWithLeftExpression:[NSExpression expressionForKeyPath:@"foo"]
+                    rightExpression:[NSExpression expressionForConstantValue:rightConstantValue]
+                           modifier:NSDirectPredicateModifier
+                               type:NSEqualToPredicateOperatorType
+                            options:0];
+    return predicate;
+}
+
+#pragma mark - String Tests
+
+- (void)testExpressionConversionString
+{
+    NSComparisonPredicate *predicate = [self equalityComparisonPredicateWithRightConstantValue:@"bar"];
+    mbgl::Value convertedValue = predicate.rightExpression.mgl_filterValue;
+    XCTAssert(convertedValue.is<std::string>() == true);
+    XCTAssertEqualObjects(@(convertedValue.get<std::string>().c_str()), @"bar");
+}
+
+#pragma mark - Boolean Tests
+
+- (void)testExpressionConversionBooleanTrue
+{
+    NSComparisonPredicate *predicate = [self equalityComparisonPredicateWithRightConstantValue:@YES];
+    mbgl::Value convertedValue = predicate.rightExpression.mgl_filterValue;
+    XCTAssert(convertedValue.is<bool>() == true);
+    XCTAssert(convertedValue.get<bool>() == true);
+}
+
+- (void)testExpressionConversionBooleanFalse
+{
+    NSComparisonPredicate *predicate = [self equalityComparisonPredicateWithRightConstantValue:@NO];
+    mbgl::Value convertedValue = predicate.rightExpression.mgl_filterValue;
+    XCTAssert(convertedValue.is<bool>() == true);
+    XCTAssert(convertedValue.get<bool>() == false);
+}
+
+#pragma mark - Floating Point Tests
+
+- (void)testExpressionConversionDouble
+{
+    NSComparisonPredicate *predicate;
+    mbgl::Value convertedValue;
+
+    predicate = [self equalityComparisonPredicateWithRightConstantValue:[NSNumber numberWithDouble:DBL_MIN]];
+    convertedValue = predicate.rightExpression.mgl_filterValue;
+    XCTAssertTrue(convertedValue.is<double>());
+    XCTAssertEqual(convertedValue.get<double>(), DBL_MIN);
+
+    predicate = [self equalityComparisonPredicateWithRightConstantValue:[NSNumber numberWithDouble:DBL_MAX]];
+    convertedValue = predicate.rightExpression.mgl_filterValue;
+    XCTAssertTrue(convertedValue.is<double>());
+    XCTAssertEqual(convertedValue.get<double>(), DBL_MAX);
+}
+
+- (void)testExpressionConversionFloat
+{
+    // Because we can't guarantee precision when using float, and because
+    // we warn the user to this effect in mgl_convertedValueWithValue:,
+    // just always pass this test.
+
+    XCTAssertTrue(YES);
```

Comment:
We should at least exercise the code path to ensure we convert to something in the right ballpark.